### PR TITLE
Adding Forest-Flow: Generating and Imputing Tabular Data via Diffusion and Flow-based Gradient-Boosted Trees

### PR DESCRIPTION
**Pull request by kilianFatras**
## What does this PR do?

This PR adds a notebook on the [Forest-Flow](https://arxiv.org/abs/2309.09968) method to TorchCFM. Forest-Flow's purpose is to generate tabular data with Flow Matching methods! We have added a notebook showing how to use XGBoost to train the vector field of I-CFM and generate tabular data. This has required the addition of a parameter 't' within the 'sample_location_and_conditional_flow' function within each class. As we have modified classes, we have also added our first test function.

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by kilianFatras** [comment]
@atong01 it seems that only the tests in runner are run and not the new one. Can you check that, please? Thank you

File: examples/tabular/README.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Forest-Flow experiment on the Iris dataset using TorchCFM
+
+This notebook is a self-contained example showing how to train the novel Forest-Flow method to generate tabular data [(Jolicoeur-Martineau et al. 2023)](https://arxiv.org/abs/2309.09968). The idea behind Forest-Flow is to **learn Independent Conditional Flow-Matching's vector field with XGBoost models** instead of neural networks. The motivation is that it is known that Forests work currently better than neural networks on Tabular data tasks. This idea comes with some difficulties, for instance how to approximate Flow Matching's loss, and this notebook shows how to do it on a minimal example. The method, its training procedure and the experiments are described in [(Jolicoeur-Martineau et al. 2023)](https://arxiv.org/abs/2309.09968). The full code can be found [here](https://github.com/SamsungSAILMontreal/ForestDiffusion).
+
+To run our jupyter notebooks, installing our package:
+
+```bash
+cd ../../
+
+# install requirements
```
Comment (atong01) [inline examples/tabular/README.md:10]: Don't need to do this, it will be done automatically with the next line

File: requirements.txt [hunk 2]
```diff
@@ -13,3 +13,9 @@ pot
 torchdiffeq
 absl-py
 clean-fid
+pytest
+
+# Forest-flow example
```
Comment (atong01) [inline requirements.txt:18]: I don't think we need these here as they are in the extra requires

File: runner-requirements.txt [hunk 3]
```diff
@@ -50,3 +50,6 @@ pot
 # --------- notebook reqs -------- #
 seaborn>=0.12.2
 pandas
+xgboost
```
Comment (atong01) [inline runner-requirements.txt:53]: Also not here

File: setup.py [hunk 4]
```diff
@@ -36,4 +36,5 @@
     long_description=readme,
     long_description_content_type="text/markdown",
     packages=find_packages(),
+    extras_require={"forest-flow": ["xgboost", "scikit-learn", "ForestDiffusion"]},
```
Comment (atong01) [inline setup.py:39]: yep perfrect

File: torchcfm/conditional_flow_matching.py [hunk 5]
```diff
@@ -164,6 +164,9 @@ def sample_location_and_conditional_flow(self, x0, x1, return_noise=False):
             represents the source minibatch
         x1 : Tensor, shape (bs, *dim)
             represents the target minibatch
+        (optionally) t : Tensor, shape (bs)
```
Comment (atong01) [inline torchcfm/conditional_flow_matching.py:167]: T here has to be batch size, but I think we should allow t as float.

Comment (atong01) [reply torchcfm/conditional_flow_matching.py:167]: Maybe better input checking would be nice, but not necessary, good for now
